lib/Sema/TypeCheckConcurrency.cpp

```diff
@@ -631,6 +631,42 @@ void swift::checkActorIsolation(const Expr *expr, const DeclContext *dc) {
             return { false, expr };
           }
         }
+
+        // Actor-isolated properties can't be passed inout to async functions.
+        // If the function being called is async, we run checkMemberReference
+        // on all of the inout parameters to ensure that they do not belong to
+        // an actor.
+        FunctionType *fnType = apply->getFn()->getType()->getAs<FunctionType>();
+        if (fnType && fnType->isAsync()) {
+          // Since we get different types for the arg list depending on
+          // whether the function is called with one parameter or multiple
+          // parameters, just collect the inout expressions so that we don't
+          // have extra special-casing
+          llvm::SmallVector<Expr *, 3> inoutArgs;
+          if (ParenExpr *arg = dyn_cast<ParenExpr>(apply->getArg())) {
+            if (InOutExpr *inoutArg = dyn_cast<InOutExpr>(arg->getSubExpr()))
+              inoutArgs.push_back(inoutArg->getSubExpr());
+          } else if (TupleExpr *argList =
+                         dyn_cast<TupleExpr>(apply->getArg())) {
+            inoutArgs.reserve(argList->getNumElements());
+            for (Expr *arg : argList->getElements()) {
+              if (InOutExpr *inoutArg = dyn_cast<InOutExpr>(arg))
+                inoutArgs.push_back(inoutArg->getSubExpr());
+            }
+          }
+          // We have a list of arguments that were passed inout to this
+          // function, check that they don't come from an actor class
+          for (Expr *arg : inoutArgs) {
+            if (LookupExpr *lookupArg = dyn_cast<LookupExpr>(arg)) {
```

Comment:
This might miss the case where the inout expression is digging into a member of an actor-isolated member, e.g.,

```swift
struct Point { var x: Int; var y: Int }
actor class A {
  var point: Point
  func f(_: inout Int) async { }
  func g() async { f(&point.x) } // should diagnose this
}
```

I wonder if instead of calling checkMemberReference directly here, we could somehow record which expressions have inout expression parents, and have checkMemberReference "look up" to find those expressions.

Reply:
Alright, when I strip off the layered `LookupExpr`s, we correctly emit an error message. Do we want the diagnostic to point at the `x` that is actually being passed, or do we want to point at the `point` since that's what belongs to the actor? Right now it looks like this, which is not wrong, but maybe confusing since it's actually passing the `x` property, not `position`:

```
/tmp/test2.swift:10:22: error: actor-isolated property 'position' cannot be passed 'inout' to asynchronous function                                                                                            
    await modifyInt(&position.x)
                     ^
```
I could probably do a note pointing at `position`, while saying that `x` is the thing that can't be passed in.

I'm doing the `LookupExpr` stripping directly in the the look over the `inoutArgs`. I don't think there would be an issue stripping off the layers of `LookupExpr`s in `checkMemberReference` since all of them do something to the effect of `lookupExpr->getBase()`. The only one that looks different is in the `SelfApplyExpr` clause where it just calls `getArg()`. I'm not sure if that's what you mean't with "having `checkMemberReference` "look up" to find the expressions"?